RollingUpgrade test fails due to missing request id

Branch: 6.x
Build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+bwc-tests/111/consoleFull

Failed task:

```
:x-pack:qa:rolling-upgrade:with-system-key:v5.6.0#oneThirdUpgradedTestCluster#wait
```

There is no reproduction line as an assertion has tripped.  The relevant failure trace is:

```
02:44:04 |    [2018-12-04T01:42:10,755][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [upgraded-node-0] fatal error in thread [elasticsearch[upgraded-node-0][refresh][T#1]], exiting
02:44:04 |    java.lang.AssertionError: Attempt to authorize action [indices:admin/refresh[s][r]] for [_xpack] without an existing request-id
02:44:04 |    	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:148) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$executeAsUser$4(ServerTransportFilter.java:184) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.maybeRun(AuthorizationUtils.java:177) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.setRunAsRoles(AuthorizationUtils.java:171) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.authorize(AuthorizationUtils.java:159) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$executeAsUser$5(ServerTransportFilter.java:187) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.core.security.SecurityContext.executeAsUser(SecurityContext.java:94) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.executeAsUser(ServerTransportFilter.java:180) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.executeAsOldVersionXPackUser(ServerTransportFilter.java:175) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$inbound$3(ServerTransportFilter.java:149) ~[?:?]
02:44:04 |    	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:60) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
02:44:04 |    	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$authenticateAsync$2(AuthenticationService.java:177) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$lookForExistingAuthentication$4(AuthenticationService.java:210) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lookForExistingAuthentication(AuthenticationService.java:221) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.authenticateAsync(AuthenticationService.java:175) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.access$000(AuthenticationService.java:134) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:104) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.inbound(ServerTransportFilter.java:130) ~[?:?]
02:44:04 |    	at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:306) ~[?:?]
02:44:04 |    	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
02:44:04 |    	at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1346) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
02:44:04 |    	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:759) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
02:44:04 |    	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
02:44:04 |    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
02:44:04 |    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
02:44:04 |    	at java.lang.Thread.run(Thread.java:834) [?:?]
```

I've also attached:

* [full build log](https://github.com/elastic/elasticsearch/files/2642898/full-build.log.zip)
* [cluster logs](https://github.com/elastic/elasticsearch/files/2642899/cluster-logs.zip) from the failing test

It seems this assertion has been added very recently so I wonder whether there are some missing edge cases? @tvernum can you please have a look at this one?

A very similar failure has occured in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+bwc-tests/112/console:

```
09:46:15 [2018-12-04T08:46:12,152][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [upgraded-node-0] fatal error in thread [elasticsearch[upgraded-node-0][management][T#1]], exiting
09:46:15 java.lang.AssertionError: Attempt to authorize action [indices:monitor/stats[n]] for [_xpack] without an existing request-id
09:46:15 	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:148) ~[?:?]
```

Another instance of this failure happened on 6.x branch: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+bwc-tests/136/console

```
[2018-12-10T23:42:03,450][INFO ][o.e.x.w.WatcherService   ] [upgraded-node-0] paused watch execution, reason [pausing watcher, cluster contains old nodes not supportin
g distributed watch execution], cancelled [0] queued tasks
[2018-12-10T23:42:03,507][INFO ][o.e.x.w.WatcherService   ] [upgraded-node-0] paused watch execution, reason [pausing watcher, cluster contains old nodes not supportin
g distributed watch execution], cancelled [0] queued tasks
[2018-12-10T23:42:03,532][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [upgraded-node-0] fatal error in thread [elasticsearch[upgraded-node-0][refresh][T#1]], e
xiting
java.lang.AssertionError: Attempt to authorize action [indices:admin/refresh[s][r]] for [_xpack] without an existing request-id
        at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:148) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$executeAsUser$4(ServerTransportFilter.java:184) ~[?:?]
        at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.maybeRun(AuthorizationUtils.java:177) ~[?:?]
        at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.setRunAsRoles(AuthorizationUtils.java:171) ~[?:?]
        at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.authorize(AuthorizationUtils.java:159) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$executeAsUser$5(ServerTransportFilter.java:187) ~[?:?]
        at org.elasticsearch.xpack.core.security.SecurityContext.executeAsUser(SecurityContext.java:94) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.executeAsUser(ServerTransportFilter.java:180) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.executeAsOldVersionXPackUser(ServerTransportFilter.java:175) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$inbound$3(ServerTransportFilter.java:149) ~[?:?]
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:60) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$authenticateAsync$2(AuthenticationService.java:177) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$lookForExistingAuthentication$4(AuthenticationService.java:210) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lookForExistingAuthentication(AuthenticationService.java:221) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.authenticateAsync(AuthenticationService.java:175) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.access$000(AuthenticationService.java:134) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:104) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.inbound(ServerTransportFilter.java:130) ~[?:?]
        at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:306) ~[?:?]
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1346) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:759) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```

Another instance of this test failure: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+bwc-tests/137/console

```
[2018-12-11T11:15:30,375][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [upgraded-node-0] fatal error in thread [elasticsearch[upgraded-node-0][management][T#1]], exiting
java.lang.AssertionError: Attempt to authorize action [indices:monitor/stats[n]] for [_xpack] without an existing request-id
        at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:148) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$executeAsUser$4(ServerTransportFilter.java:184) ~[?:?]
        at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.maybeRun(AuthorizationUtils.java:177) ~[?:?]
        at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.setRunAsRoles(AuthorizationUtils.java:171) ~[?:?]
        at org.elasticsearch.xpack.security.authz.AuthorizationUtils$AsyncAuthorizer.authorize(AuthorizationUtils.java:159) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$executeAsUser$5(ServerTransportFilter.java:187) ~[?:?]
        at org.elasticsearch.xpack.core.security.SecurityContext.executeAsUser(SecurityContext.java:94) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.executeAsUser(ServerTransportFilter.java:180) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.executeAsOldVersionXPackUser(ServerTransportFilter.java:175) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$inbound$3(ServerTransportFilter.java:149) ~[?:?]
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:60) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$authenticateAsync$2(AuthenticationService.java:177) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$lookForExistingAuthentication$4(AuthenticationService.java:210) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lookForExistingAuthentication(AuthenticationService.java:221) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.authenticateAsync(AuthenticationService.java:175) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.access$000(AuthenticationService.java:134) ~[?:?]
        at org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:104) ~[?:?]
        at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.inbound(ServerTransportFilter.java:130) ~[?:?]
        at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:306) ~[?:?]
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1346) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:759) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-6.6.0-SNAPSHOT.jar:6.6.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```

And another one, similar but on a different action: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+bwc-tests/141/console

```
09:03:03  |    [2018-12-12T07:01:09,210][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [upgraded-node-0] fatal error in thread [elasticsearch[upgraded-node-0][management][T#1]], exiting
09:03:03  |    java.lang.AssertionError: Attempt to authorize action [cluster:monitor/xpack/ml/job/stats/get[n]] for [_xpack] without an existing request-id
09:03:03  |    	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:148) ~[?:?]
```

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/BwcXPackUser.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/BwcXPackUser.java
@@ -1,0 +1,40 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.security.user;
+
+/**
+ * The representation of the XPackUser prior to version 5.6.1. This class should only be used for BWC purposes!
+ */
+@Deprecated
+public final class BwcXPackUser extends User {
+
+    public static final String NAME = XPackUser.NAME;
+    public static final String ROLE_NAME = "superuser";
+    public static final BwcXPackUser INSTANCE = new BwcXPackUser();
+
+    private BwcXPackUser() {
+        super(NAME, ROLE_NAME);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return INSTANCE == o;
+    }
+
+    @Override
+    public int hashCode() {
+        return System.identityHashCode(this);
+    }
+
+    public static boolean is(User user) {
+        return INSTANCE.equals(user);
+    }
+
+    public static boolean is(String principal) {
+        return NAME.equals(principal);
+    }
+}

--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
@@ -6,6 +6,7 @@
 package org.elasticsearch.xpack.security.authz;
 
 import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.CompositeIndicesRequest;
 import org.elasticsearch.action.DocWriteRequest;
@@ -57,6 +58,7 @@ import org.elasticsearch.xpack.core.security.authz.privilege.IndexPrivilege;
 import org.elasticsearch.xpack.core.security.authz.store.ReservedRolesStore;
 import org.elasticsearch.xpack.core.security.support.Automatons;
 import org.elasticsearch.xpack.core.security.user.AnonymousUser;
+import org.elasticsearch.xpack.core.security.user.BwcXPackUser;
 import org.elasticsearch.xpack.core.security.user.SystemUser;
 import org.elasticsearch.xpack.core.security.user.User;
 import org.elasticsearch.xpack.core.security.user.XPackSecurityUser;
@@ -139,7 +141,7 @@ public class AuthorizationService extends AbstractComponent {
         if (auditId == null) {
             // We would like to assert that there is an existing request-id, but if this is a system action, then that might not be
             // true because the request-id is generated during authentication
-            if (isInternalUser(authentication.getUser()) != false) {
+            if (isInternalUser(authentication) != false) {
                 auditId = AuditUtil.getOrGenerateRequestId(threadContext);
             } else {
                 auditTrail.tamperedRequest(null, authentication.getUser(), action, request);
@@ -353,8 +355,10 @@ public class AuthorizationService extends AbstractComponent {
         auditTrail.accessGranted(auditId, authentication, action, request, permission.names());
     }
 
-    private boolean isInternalUser(User user) {
-        return SystemUser.is(user) || XPackUser.is(user) || XPackSecurityUser.is(user);
+    private boolean isInternalUser(Authentication authentication) {
+        final User user = authentication.getUser();
+        return SystemUser.is(user) || XPackUser.is(user) || XPackSecurityUser.is(user) ||
+            (BwcXPackUser.is(user) && authentication.getVersion().before(Version.V_5_6_1));
     }
 
     private boolean hasSecurityIndexAccess(IndicesAccessControl indicesAccessControl) {
@@ -469,6 +473,10 @@ public class AuthorizationService extends AbstractComponent {
             return;
         }
         if (XPackSecurityUser.is(user)) {
+            roleActionListener.onResponse(ReservedRolesStore.SUPERUSER_ROLE);
+            return;
+        }
+        if (BwcXPackUser.is(user)) {
             roleActionListener.onResponse(ReservedRolesStore.SUPERUSER_ROLE);
             return;
         }

--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/ServerTransportFilter.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/ServerTransportFilter.java
@@ -27,6 +27,7 @@ import org.elasticsearch.transport.TransportService;
 import org.elasticsearch.transport.netty4.Netty4TcpChannel;
 import org.elasticsearch.xpack.core.security.SecurityContext;
 import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.user.BwcXPackUser;
 import org.elasticsearch.xpack.core.security.user.KibanaUser;
 import org.elasticsearch.xpack.core.security.user.SystemUser;
 import org.elasticsearch.xpack.core.security.user.User;
@@ -171,8 +172,7 @@ public interface ServerTransportFilter {
 
         private void executeAsOldVersionXPackUser(String securityAction, TransportRequest request, TransportChannel transportChannel,
                                                   ActionListener<Void> listener) {
-            final User xpackUser = new User(XPackUser.NAME, "superuser");
-            executeAsUser(xpackUser, securityAction, request, transportChannel, listener);
+            executeAsUser(BwcXPackUser.INSTANCE, securityAction, request, transportChannel, listener);
         }
 
         private void executeAsUser(User user, String securityAction, TransportRequest request, TransportChannel transportChannel,

--- a/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/ServerTransportFilterTests.java
+++ b/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/ServerTransportFilterTests.java
@@ -27,6 +27,7 @@ import org.elasticsearch.xpack.core.security.authc.Authentication;
 import org.elasticsearch.xpack.core.security.authc.Authentication.RealmRef;
 import org.elasticsearch.xpack.core.security.authz.permission.Role;
 import org.elasticsearch.xpack.core.security.authz.store.ReservedRolesStore;
+import org.elasticsearch.xpack.core.security.user.BwcXPackUser;
 import org.elasticsearch.xpack.core.security.user.KibanaUser;
 import org.elasticsearch.xpack.core.security.user.SystemUser;
 import org.elasticsearch.xpack.core.security.user.User;
@@ -44,6 +45,7 @@ import static org.elasticsearch.xpack.core.security.support.Exceptions.authentic
 import static org.elasticsearch.xpack.core.security.support.Exceptions.authorizationError;
 import static org.hamcrest.Matchers.arrayContaining;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.sameInstance;
 import static org.mockito.Matchers.any;
 import static org.mockito.Matchers.eq;
 import static org.mockito.Matchers.isA;
@@ -274,31 +276,31 @@ public class ServerTransportFilterTests extends ESTestCase {
             callback.onResponse(authentication);
             return Void.TYPE;
         }).when(authcService).authenticate(eq("_action"), eq(request), eq((User) null), any(ActionListener.class));
-        AtomicReference<String[]> rolesRef = new AtomicReference<>();
+        AtomicReference<User> userRef = new AtomicReference<>();
         final Role empty = Role.EMPTY;
         doAnswer((i) -> {
             ActionListener callback =
                     (ActionListener) i.getArguments()[1];
-            rolesRef.set(((User) i.getArguments()[0]).roles());
+            userRef.set((User) i.getArguments()[0]);
             callback.onResponse(empty);
             return Void.TYPE;
         }).when(authzService).roles(any(User.class), any(ActionListener.class));
         ServerTransportFilter filter = getClientOrNodeFilter();
         PlainActionFuture<Void> future = new PlainActionFuture<>();
         filter.inbound("_action", request, channel, future);
-        assertNotNull(rolesRef.get());
-        assertThat(rolesRef.get(), arrayContaining("superuser"));
+        assertThat(userRef.get(), sameInstance(BwcXPackUser.INSTANCE));
+        assertThat(userRef.get().roles(), arrayContaining("superuser"));
 
         // test with a version that doesn't need changing
         filter = getClientOrNodeFilter();
-        rolesRef.set(null);
+        userRef.set(null);
         user = XPackUser.INSTANCE;
         when(authentication.getUser()).thenReturn(user);
         when(authentication.getVersion()).thenReturn(Version.V_5_6_1);
         future = new PlainActionFuture<>();
         filter.inbound("_action", request, channel, future);
-        assertNotNull(rolesRef.get());
-        assertThat(rolesRef.get(), arrayContaining(XPackUser.ROLE_NAME));
+        assertThat(userRef.get(), sameInstance(XPackUser.INSTANCE));
+        assertThat(userRef.get().roles(), arrayContaining(XPackUser.ROLE_NAME));
     }
 
     private ServerTransportFilter getClientOrNodeFilter() throws IOException {

